pandas/core/arrays/datetimes.py

```diff
@@ -326,6 +344,24 @@ def _generate_range(cls, start, end, periods, freq, tz=None,
 
         return cls._simple_new(index.asi8, freq=freq, tz=tz)
 
+    # -----------------------------------------------------------------
+    # DatetimeLike Interface
+
+    def _unbox_scalar(self, value):
+        assert isinstance(value, self._scalar_type), value
+        return value.value
+
+    def _scalar_from_string(self, value):
+        assert isinstance(value, self._scalar_type), value
+        return Timestamp(value)
+
+    def _check_compatible_with(self, other):
+        # TODO: verify this.
+        if not timezones.tz_compare(self.tz, other.tz):
+            raise IncompatibleTimeZoneError(
```

Comment:
Do people just want a ValueError here? Some context on why I did this (and why it didn't end up mattering). See DatetimeTZBlock.setitem for more.

`Series.__setitem__` allows setting values that will cast the type if the *array* (e.g. setting a Timestamp with a different TZ, or even non timestamps). But `DatetimeArray` will always be limited to a single tz for all the values.


In `DatetimeTZBlock.setitem`, I had hoped we could do the array `__setitem__` and catch just `IncompatibleTimeZoneError`, convert to object, and retry. But `Series.__setitem__` allows setting pretty much, so we have to catch type errors to do things like

```python
In [5]: tzser = pd.Series(pd.date_range("2000", periods=2, tz="UTC"))

In [6]: tzser[0] = 'a'

In [7]: tzser
Out[7]:
0                            a
1    2000-01-02 00:00:00+00:00
dtype: object
```

Reply:
> See DatetimeTZBlock.setitem for more

Related to the pickle discussion in #24074, each of the setitem variants needs to invalidate the `freq` attribute.  I'll be opening a PR shortly that does this in DatetimeTZBlock._try_coerce_result, not sure where else it needs to be done.